compiler-core/src/language_server/tests/action.rs

```diff
@@ -0,0 +1,225 @@
+use crate::line_numbers::LineNumbers;
+use lsp_types::{
+    CodeActionContext, CodeActionParams, Diagnostic, PartialResultParams, Position, Range,
+    TextDocumentIdentifier, Url, WorkDoneProgressParams, WorkspaceEdit,
+};
+
+use super::*;
+
+fn remove_unused_action(src: &str) -> String {
+    do_remove_unused_action(src).unwrap_or_else(|| src.into())
+}
+
+fn do_remove_unused_action(src: &str) -> Option<String> {
+    let io = LanguageServerTestIO::new();
+    let mut engine = setup_engine(&io);
+
+    // inject stdlib stubs
+    _ = io.src_module("list", "");
+    _ = io.src_module(
+        "result",
+        "pub fn is_ok() {}\npub fn is_err() {}\npub fn all() {}",
+    );
+    _ = io.src_module("map", "pub type Map(key, value)\npub fn delete() {}");
+    _ = io.src_module("option", "");
+
+    _ = io.src_module("app", src);
+    engine.compile_please().result.expect("compiled");
+
+    let path = Utf8PathBuf::from(if cfg!(target_family = "windows") {
+        r#"\\?\C:\src\app.gleam"#
+    } else {
+        "/src/app.gleam"
+    });
+
+    let url = Url::from_file_path(path).unwrap();
+
+    let diag = Diagnostic {
+        range: Range::new(Position::new(0, 0), Position::new(0, 0)),
+        severity: None,
+        code: None,
+        code_description: None,
+        source: None,
+        message: "Unused".to_string(),
+        related_information: None,
+        tags: None,
+        data: None,
+    };
+
+    let params = CodeActionParams {
+        text_document: TextDocumentIdentifier::new(url),
+        context: CodeActionContext {
+            diagnostics: vec![diag],
+            only: None,
+        },
+        range: Range::new(Position::new(0, 0), Position::new(0, 0)),
+        work_done_progress_params: WorkDoneProgressParams {
+            work_done_token: None,
+        },
+        partial_result_params: PartialResultParams {
+            partial_result_token: None,
+        },
+    };
+    let response = engine.action(params).result.unwrap();
+    if let Some(action) = response {
+        let action = action.iter().next()?;
+        Some(apply_code_action(src, action))
+    } else {
+        None
+    }
+}
+
+fn apply_code_action(src: &str, action: &lsp_types::CodeAction) -> String {
+    let mut result = src.to_string();
+    let line_numbers = LineNumbers::new(src);
+    let mut offset = 0;
+    match &action.edit {
+        Some(WorkspaceEdit { changes, .. }) => match changes {
+            Some(changes) => {
+                for change in changes.values() {
```

Comment:
Could the code in this branch be extracted into another function please? To split out deciding what needs to be done from how it is done.